lib/pbench/agent/tool_meister.py

```diff
@@ -110,20 +127,112 @@ def __init__(
         assert (
             pbench_install_dir is not None
         ), "Logic bomb!  no installation directory provided!"
+        assert (
+            pbench_install_dir.is_dir
+        ), f"pbench installation directory does not exist: {pbench_install_dir}"
         self.pbench_install_dir = pbench_install_dir
+        assert tool_dir is not None, "Logic bomb!  no tool directory provided!"
+        assert tool_dir.is_dir, f"tool directory does not exist: {tool_dir}"
         self.tool_dir = tool_dir
-        self.start_process = None
-        self.stop_process = None
 
-    def _check_no_processes(self):
-        if self.start_process is not None:
-            raise ToolException(
-                f"Tool({self.name}) has an unexpected start process running"
+    def install(self):
+        raise NotImplementedError(
+            f"{self.__class__.name} does not implement the install method"
+        )
+
+    def start(self):
+        raise NotImplementedError(
+            f"{self.__class__.name} does not implement the start method"
+        )
+
+    def stop(self):
+        raise NotImplementedError(
+            f"{self.__class__.name} does not implement the stop method"
+        )
+
+    def wait(self):
+        raise NotImplementedError(
+            f"{self.__class__.name} does not implement the wait method"
+        )
+
+    def _wait_for_process(self, process):
+        """Generic method to wait for a given process to stop after 30
+        seconds, emitting a message every 5 seconds in between.
+
+        Returns the process return code on success, or None if the process
+        failed to exit.
+        """
+        count = 6
+        sts = None
+        while count > 0 and sts is None:
+            try:
+                sts = process.wait(timeout=5)
+            except subprocess.TimeoutExpired:
+                count -= 1
+                if count > 0:
+                    self.logger.debug(
+                        "%s tool %s has not stopped after %d seconds",
+                        self._tool_type,
+                        self.name,
+                        5 * (6 - count),
+                    )
+        return sts
+
+    def _wait_for_process_w_kill(self, process, ctx_name):
+        """Generic method of waiting for a given process, killing the process
+        if the initial wait failed, waiting a second time for the kill to take
+        effect.
+
+        Always returns None
+        """
+        self.logger.info(
+            "Waiting for %s tool %s %s process", self._tool_type, self.name, ctx_name
+        )
+        # We wait for the {ctx_name} process to finish first ...
+        sts = self._wait_for_process(process)
+        if sts is None:
+            # The {ctx_name} process did not terminate gracefully after 30 seconds,
+            # so we bring out the big guns ...
+            process.kill()
+            self.logger.error(
+                "Killed un-responsive %s tool %s %s process after" " 30 seconds",
```

Comment:
This message should omit the "after 30 seconds", since this code doesn't know how long `_wait_for_process()` waited.  (Also, this gets rid of the embedded `" "` break.)

(Of course, having `_wait_for_process()` take the overall timeout as a parameter might be a good thing....)